bpo-42128: Add Pattern Matching to What's New (#24667)
python/cpython: Doc/whatsnew/3.10.rst

(Contributed by Ken Jin in :issue:`41559`.)

PEP 634: Structural Pattern Matching
------------------------------------

Structural pattern matching has been added in the form of a *match statement*,

> a data object (sequences, mappings, primitive data types as well as class

Reviewer: Is a missing here? E.g. "containing" or "which consists of"?

instances) and *case statements* of patterns with associated actions. Pattern
matching enables programs to extract information from complex data types,
branch on the structure of data, and apply specific actions based on different
forms of data.

Syntax and operations
~~~~~~~~~~~~~~~~~~~~~

The generic syntax of pattern matching is::

>     def function(subject):

Reviewer: Why wrap it in a dummy function?

Author: No reason, I was doing several drafts and this was part of another example. Removed the wrap.

>         match {subject}:

Reviewer: Why the curly braces? (If this is "meta variables" as I suspect I'd think we usually use `<angular brackets>` over `{curlies}`. Plus, `subject` is the parameter of the function so maybe it shouldn't be a meta variable at all?)

            case {pattern_1}:
                {action_1}
            case {pattern_2}:
                {action_2}
            case {pattern_3}:
                {action_3}
            case _:
                {action_wildcard}

A match statement takes an expression and compares its value to successive
patterns given as one or more case blocks.  Specifically, pattern matching

> operates by::

Reviewer: ```suggestion
operates by:
```

Reviewer: This still needs to be applied.

    1. taking a data object ``subject`` with type and shape

>     2. substituting the ``subject`` into a ``match`` statement

Reviewer 2: I'm not sure what "substitution" this explanation refers to. Are you trying to separate: 1. the expression on the map statement is evaluated 2. the resulting value is considered the "subject" ?

Author: @dmoisset Thanks for the suggestion. I've done a second pass which hopefully reads better. I was, perhaps ineloquently, trying to reinforce the importance of data, its type, and its shape so folks could view this as more than a simple switch statement with literals. Let me know if you have more suggestions.

>     3. comparing the match subject with each pattern in a ``case`` statement

Reviewer: Maybe drop "match"?
```suggestion
    3. comparing the subject with each pattern in a ``case`` statement
```

       from top to bottom until a match is confirmed. If an exact match is not
       confirmed, the last case, a wildcard ``_``, will be used as the matching

>        case.

Reviewer: Hm, this suggests that there should always be a wildcard case at the end. Unless you are talking specifically about the above example, but in that case I would open with "This match statement takes" instead of "A match statement takes". Alternatively, you could perhaps add something like "If no wildcard pattern is present, and an exact match is not confirmed, the entire match block is a no-op." (Though maybe the whole thing about the wildcard can be moved down to a new item 5.)

>     4. executing the actions associated with the pattern of the confirmed

Reviewer: ```suggestion
    4. executing the action associated with the pattern of the confirmed
```

       match

Declarative approach
~~~~~~~~~~~~~~~~~~~~

Readers may be aware of pattern matching through the simple example of matching
a subject (data object) to a literal (pattern) with the switch statement found
in C, Java or JavaScript (and many other languages). Often the switch statement
is used for comparison of an object/expression with case statements containing

> literals. 

Reviewer 3: ```suggestion
literals.
```

> More powerful examples of pattern matching can be found in languages, such as
> Elixir. With structural pattern matching, the approach is "declarative" and

Reviewer: ```suggestion
More powerful examples of pattern matching can be found in languages such as
Scala. With structural pattern matching, the approach is "declarative" and
```
(I really hate how matching looks like in Elixir, but I was truly inspired when I saw how Scala did it. There's at least one Scala dev who will appreciate this call out.)

Author: I've kept Elixir and added Scala before it. There are some folks who really like the Elixir pattern matching and will be happy with the PEP addition. With you on how it looks

> explicitly states the conditions (the patterns) for data to match. 

Reviewer 3: ```suggestion
explicitly states the conditions (the patterns) for data to match.
```

While an "imperative" series of instructions using nested "if" statements
could be used to accomplish something similar to structural pattern matching,

> it is less clear than the "declarative" approach. Instead the "declarative"

Reviewer: ```suggestion
it is less clear than the "declarative" approach. Instead, the "declarative"
```

approach states the conditions to meet for a match and is more readable through
its explicit patterns. While structural pattern matching can be used in its
simplest form comparing a variable to a literal in a case statement, its

> true value for Python lies in its handling of data's type and shape.

Reviewer: ```suggestion
true value for Python lies in its handling of the subject's type and shape.
```

Simple pattern: match to a literal
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Let's look at this example as pattern matching in its simplest form: a value, 

Reviewer 3: ```suggestion
Let's look at this example as pattern matching in its simplest form: a value,
```

the subject, being matched to several literals, the patterns. In the example

> below, ``status`` is the subject of the match statement. The patterns are 

Reviewer 3: ```suggestion
below, ``status`` is the subject of the match statement. The patterns are
```

            print(f"X={x}")
        case (x, y):
            print(f"X={x}, Y={y}")
        case _:
            raise ValueError("Not a point")

The first pattern has two literals, ``(0, 0)``, and may be thought of as an
extension of the literal pattern shown above. The next two patterns combine a
literal and a variable, and the variable *binds* a value from the subject
(``point``).  The fourth pattern captures two values, which makes it
conceptually similar to the unpacking assignment ``(x, y) = point``.

Patterns and classes
~~~~~~~~~~~~~~~~~~~~

If you are using classes to structure your data, you can use as a pattern
the class name followed by an argument list resembling a constructor. This
pattern has the ability to capture class attributes into variables::

    class Point:
        x: int
        y: int

    def location(point):
        match point:
            case Point(x=0, y=0):
                print("Origin is the point's location.")
            case Point(x=0, y=y):
                print(f"Y={y} and the point is on the y-axis.")
            case Point(x=x, y=0):
                print(f"X={x} and the point is on the x-axis.")
            case Point():

>                 print("The point is located somewhere else on the graph.")

Reviewer: ```suggestion
                print("The point is located somewhere else on the plane.")
```

list of points, it could be matched like this::

    match points:
        case []:
            print("No points in the list.")
        case [Point(0, 0)]:
            print("The origin is the only point in the list.")
        case [Point(x, y)]:
            print(f"A single point {x}, {y} is in the list.")
        case [Point(0, y1), Point(0, y2)]:
            print(f"Two points on the Y axis at {y1}, {y2} are in the list.")
        case _:
            print("Something else is found in the list.")

Guard
~~~~~

We can add an ``if`` clause to a pattern, known as a "guard".  If the
guard is false, ``match`` goes on to try the next case block.  Note
that value capture happens before the guard is evaluated::

    match point:
        case Point(x, y) if x == y:
            print(f"The point is located on the diagonal Y=X at {x}.")
        case Point(x, y):
            print(f"Point is not on the diagonal.")

Other Key Features
~~~~~~~~~~~~~~~~~~

Several other key features:

- Like unpacking assignments, tuple and list patterns have exactly the
  same meaning and actually match arbitrary sequences. Technically,
  the subject must be an instance of ``collections.abc.Sequence``.

>   Therefore, an important exception is that patterns don't match iterators 

Reviewer 3: ```suggestion
  Therefore, an important exception is that patterns don't match iterators
```

>   Therefore, an important exception is that patterns don't match iterators 
>   or strings.

Reviewer: ```suggestion
  Therefore, an important exception is that patterns don't match iterators.
  Also, to prevent a common mistake, sequence patterns don't match strings.
```

- Sequence patterns support wildcards: ``[x, y, *rest]`` and ``(x, y,
  *rest)`` work similar to wildcards in unpacking assignments.  The
  name after ``*`` may also be ``_``, so ``(x, y, *_)`` matches a sequence
  of at least two items without binding the remaining items.

- Mapping patterns: ``{"bandwidth": b, "latency": l}`` captures the
  ``"bandwidth"`` and ``"latency"`` values from a dict.  Unlike sequence
  patterns, extra keys are ignored.  A wildcard ``**rest`` is also
  supported.  (But ``**_`` would be redundant, so it not allowed.)

- Subpatterns may be captured using the ``as`` keyword::

>       case (Point(x1, y1), Point(x2, y2) as p2): ...

Reviewer: This could use some clarification: this binds x1, y1, x2, y2 like you'd expect without the `as` clause, and p2 to the entire second item of the subject.

- Most literals are compared by equality. However, the singletons ``True``,
  ``False`` and ``None`` are compared by identity.

- Named constants may be used in patterns.  These named constants must be
  dotted names to prevent the constant from being interpreted as a capture

>    variable::

Reviewer: ```suggestion
  variable::
```

------------------------------------

Structural pattern matching has been added in the form of a *match statement*,
a data object (which consists of sequences, mappings, primitive data types as well as class

> instances) and *case statements* of patterns with associated actions. Pattern

Reviewer: I still think this is confusing -- to me it suggests that "a data object" was added to the language. Does this refer to the (syntactic) concept of a *pattern*? How about the following rewrite?
```suggestion
and *case statements* of patterns with associated actions. Patterns 
consists of sequences, mappings, primitive data types as well as class instances. Pattern
```

Author: Good point re: addition. Applied your suggestion.
